src/tls/openssl.c

```diff
@@ -353,13 +353,14 @@ static int tls_net_read(struct flb_upstream_conn *u_conn,
     ctx = session->parent;
     pthread_mutex_lock(&ctx->mutex);
 
+    ERR_clear_error();
     ret = SSL_read(session->ssl, buf, len);
     if (ret <= 0) {
         ret = SSL_get_error(session->ssl, ret);
         if (ret == SSL_ERROR_WANT_READ) {
             ret = FLB_TLS_WANT_READ;
         }
-        else if (ret < 0) {
```

Comment:
here is a thing (and I still need to clarify more): the error might describe that "wants to read or write" in the case of "SSL_ERROR_WANT_WRITE", by definition:

```
#define SSL_ERROR_WANT_WRITE 3
```

that is not an error and removing the ```< 0``` will make that case to fail. Not sure about the impact of that change.

Reply:
I understand that. However it is already a bug. tls_net_read is returning 3 today for `SSL_ERROR_WANT_WRITE` which is worse because it makes the http client adds 3 garbage bytes to its response buffer. It doesn't log an error but it will corrupt memory later on.

Reply:
Added the change to handle SSL_ERROR_WANT_WRITE.